client/peerset/src/lib.rs

```diff
@@ -216,159 +257,178 @@ impl Peerset {
+		self.data.add_no_slot_node(set_id.0, peer_id);
+		self.alloc_slots();
 	}
 
-	fn on_set_reserved_only(&mut self, reserved_only: bool) {
-		self.reserved_only = reserved_only;
+	fn on_remove_reserved_peer(&mut self, set_id: SetId, peer_id: PeerId) {
+		if !self.reserved_nodes[set_id.0].0.remove(&peer_id) {
+			return;
+		}
 
-		if self.reserved_only {
-			// Disconnect all the nodes that aren't reserved.
-			for peer_id in self.data.connected_peers().cloned().collect::<Vec<_>>().into_iter() {
-				if self.priority_groups.get(RESERVED_NODES).map_or(false, |g| g.contains(&peer_id)) {
-					continue;
-				}
+		self.data.remove_no_slot_node(set_id.0, &peer_id);
 
-				let peer = self.data.peer(&peer_id).into_connected()
-					.expect("We are enumerating connected peers, therefore the peer is connected; qed");
-				peer.disconnect();
-				self.message_queue.push_back(Message::Drop(peer_id));
-			}
+		// Nothing more to do if not in reserved-only mode.
+		if !self.reserved_nodes[set_id.0].1 {
+			return;
+		}
 
-		} else {
-			self.alloc_slots();
+		// If, however, the peerset is in reserved-only mode, then the removed node needs to be
+		// disconnected.
+		if let peersstate::Peer::Connected(peer) = self.data.peer(set_id.0, &peer_id) {
+			peer.disconnect();
+			self.message_queue.push_back(Message::Drop {
+				set_id,
+				peer_id,
+			});
 		}
 	}
 
-	fn on_set_priority_group(&mut self, group_id: &str, peers: HashSet<PeerId>) {
+	fn on_set_reserved_peers(&mut self, set_id: SetId, peer_ids: HashSet<PeerId>) {
 		// Determine the difference between the current group and the new list.
 		let (to_insert, to_remove) = {
-			let current_group = self.priority_groups.entry(group_id.to_owned()).or_default();
-			let to_insert = peers.difference(current_group)
+			let to_insert = peer_ids.difference(&self.reserved_nodes[set_id.0].0)
 				.cloned().collect::<Vec<_>>();
-			let to_remove = current_group.difference(&peers)
+			let to_remove = self.reserved_nodes[set_id.0].0.difference(&peer_ids)
 				.cloned().collect::<Vec<_>>();
 			(to_insert, to_remove)
 		};
 
-		// Enumerate elements in `peers` not in `current_group`.
-		for peer_id in &to_insert {
-			// We don't call `on_add_to_priority_group` here in order to avoid calling
-			// `alloc_slots` all the time.
-			self.priority_groups.entry(group_id.to_owned()).or_default().insert(peer_id.clone());
-			self.data.add_no_slot_node(peer_id.clone());
+		for node in to_insert {
+			self.on_add_reserved_peer(set_id, node);
 		}
 
-		// Enumerate elements in `current_group` not in `peers`.
-		for peer in to_remove {
-			self.on_remove_from_priority_group(group_id, peer);
+		for node in to_remove {
+			self.on_remove_reserved_peer(set_id, node);
 		}
+	}
+
+	fn on_set_reserved_only(&mut self, set_id: SetId, reserved_only: bool) {
+		self.reserved_nodes[set_id.0].1 = reserved_only;
+
+		if reserved_only {
+			// Disconnect all the nodes that aren't reserved.
+			for set_index in 0..self.data.num_sets() {
```

Comment:
I am a bit confused by the fact that `on_set_reserved_only` is specific to a `SetId`, but still the logic below disconnects peers of all sets in case they are not reserved. Why is that?

Reply:
Fixed